File: FarmaciaERP/Backend/src/main/java/FarmaciaERP/Domain/Entities/Medicamento.java
package FarmaciaERP.Domain.Entities;

import FarmaciaERP.Domain.Enums.MedicamentoCategoria;
import lombok.Getter;
import lombok.Setter;

import java.time.LocalDate;

@Getter
@Setter
public class Medicamento {
    private int id;
    private String nombre;
    private String presentacion;
    private double precio;
    private int stock;
    private MedicamentoCategoria categoria;
    private LocalDate fechaVencimiento;

    public Medicamento() {}

    public Medicamento(int id, String nombre, String presentacion, double precio,
                       int stock, MedicamentoCategoria categoria, LocalDate fechaVencimiento) {
        this.id = id;
        this.nombre = nombre;
        this.presentacion = presentacion;
        this.precio = precio;
        this.stock = stock;
        this.categoria = categoria;
        this.fechaVencimiento = fechaVencimiento;
    }

    public boolean estaVencido() {
        return fechaVencimiento != null && fechaVencimiento.isBefore(LocalDate.now());
    }
    public boolean requiereReceta(){
        return  categoria.requiereReceta();
    }
    public boolean estaDisponible(){
        return stock > 0 && !estaVencido();
    }

}

File: FarmaciaERP/frontend-react/src/domain/models/Medicamento.js
export class Medicamento {
  constructor({ id, nombre, presentacion, precio, stock, categoria, fechaVencimiento }) {
    this.id = id;
    this.nombre = nombre;
    this.presentacion = presentacion ?? '';
    this.precio = Number(precio ?? 0);
    this.stock = Number(stock ?? 0);
    this.categoria = categoria ?? 'VENTA_LIBRE';
    this.fechaVencimiento = fechaVencimiento ?? null;
  }

  get requiereReceta() {
    return this.categoria === 'RECETA_SIMPLE' || this.categoria === 'RECETA_RETENIDA' || this.categoria === 'CONTROLADO';
  }

  get estaVencido() {
    if (!this.fechaVencimiento) return false;
    return new Date(this.fechaVencimiento) < new Date();
  }

  get estaDisponible() {
    return this.stock > 0 && !this.estaVencido;
  }

  get stockBajo() {
    return this.stock > 0 && this.stock <= 10;
  }

  static fromApi(raw) {
    return new Medicamento({
      id: raw.id,
      nombre: raw.nombre,
      presentacion: raw.presentacion,
      precio: raw.precio,
      stock: raw.stock,
      categoria: raw.categoria,
      fechaVencimiento: raw.fechaVencimiento,
    });
  }
}
